### PR TITLE
fix binary compat problem for inref struct pointers

### PR DESCRIPTION
**Pull request by dsyme**
This fixes the cause of https://github.com/Microsoft/visualfsharp/issues/5369

F# 4.5 changed "this" pointers on struct instance members to have type "inref" for the purposes of checking the implementation.  However, the presence of this `inref` caused a change in the F# metadata for existing code.  

There is no need for external users of these members to see the use of the "inref" pointer. So this change rewrites the uses of "inref" pointers occuring in the "this" position for instance members on structs so that they are simply "byref" pointers on assembly boundaries.

You must build proto to apply this fix to FSharp.Core:

    .\build net40 debug proto

So far tested manually by having:

**lib.fs**
```
module Lib

[<Struct>]
type S(x:int) = 
    member this.M() = x + 1
    member this.P = x + 2
```

**app.fs**
```

module App

open Lib

let s = S()

let y = s.M()
let y2 = s.P


let a = async { return 0 }
```

and then 
```
# Make sure we have updated compiler and FSHarp.Core.dll
.\build net40 debug proto

# Compile a component using new compiler
debug\net40\bin\fsc -a lib.fs 

# Reference the resulting component using an **old** compiler, also reference the new FSharp.Core from old compiler
packages\FSharp.Compiler.Tools.4.1.27\tools\fsc.exe --noframework -r:lib.dll app.fs -r:debug\net40\bin\FSharp.Core.dll  -r:c:\Windows\Microsoft.NET\Framework\v4.0.30319\mscorlib.dll
```

With this in place, using the old compiler no longer gives a warning.

### CODE REVIEW

**Comment by Pilchie** [comment]
Thanks @dsyme !

**Comment by cartermp** [comment]
Just to confirm, this still prevents evil struct replacement yes?


On Thu, Jul 26, 2018, at 09:14, Kevin Pilch wrote:
> Thanks @dsyme[1] !


> — You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub[2], or mute the
> thread[3].

Links:

  1. https://github.com/dsyme
  2. https://github.com/Microsoft/visualfsharp/pull/5388#issuecomment-408151800

**Comment by dsyme** [comment]
@Pilchie Mea culpa for causing the problem.

From internal discussion, here's the plan:
•	The bug is about pre dev15.8-tooling consuming dev15.8-compiled components, including FSharp.Core
•	The fix will be in the F# compiler to make sure F#-specific metadata produced for "this" members on structs is identical to what it was before
•	We will test that you can use dev15.8 tooling to make a user DLL that no longer gives a warning when consumed by pre-dev 15.8 tooling.
•	Part of rolling out the fix will involve recompiling and republishing a new signed FSharp.Core, presumably package number 4.5.2
•	We will test that you can use that FSharp.Core with pre dev 15.8 tooling

I also proposed this:
 
•	To avoid this in the future we will add some kind of automated testing to make sure the F#-specific metadata being produced for FSharp.Core is “pinned down” as rigidly as the .NET-specific metadata and doesn’t change.  We will do this by “printing” the F#-specific metadata to a baseline format of some kind.
 
•	Ideally we will also add a smoke test of consuming more recently produced FSharp.Core with pre dev 15.8-tooling, e.g. compiling some code with old versions of the fsc.exe in FSharp.Copiler.Tools should be adequate.

**Comment by dsyme** [comment]
> Just to confirm, this still prevents evil struct replacement yes?

Yes, it should.  This only affects the persisted metadata across assembly boundaries

File: src/fsharp/TastPickle.fs [hunk 1]
```diff
@@ -1277,7 +1287,13 @@ let p_nonlocal_val_ref (nlv:NonLocalValOrMemberRef) st =
     p_bool pkey.MemberIsOverride st 
     p_string pkey.LogicalName st 
     p_int pkey.TotalArgCount st 
+    let isStructThisArgPos = 
+        match key.TypeForLinkage with
```
Comment (TIHan) [inline src/fsharp/TastPickle.fs]: What does TypeForLinkage here mean?

File: src/fsharp/TastPickle.fs [hunk 2]
```diff
@@ -1268,6 +1271,13 @@ let p_tys = (p_list p_ty)
 
 let fill_p_attribs,p_attribs = p_hole()
 
+// In F# 4.5, the type of the "this" pointer for structs is considered to be inref for the purposes of checking the implementation
+// of the struct.  However for backwards compat reaons we can't serialize this as the type.
+let checkForInRefStructThisArg st ty = 
+    let g = st.oglobals
```
Comment (TIHan) [inline src/fsharp/TastPickle.fs]: What is `st.oglobals`?

File: src/fsharp/TastPickle.fs [hunk 3]
```diff
@@ -1539,6 +1555,7 @@ let u_tyar_specs = (u_list u_tyar_spec)
 
 let _ = fill_p_ty (fun ty st ->
     let ty = stripTyparEqns ty
+    let ty = if isInByrefTy st.oglobals ty && st.isStructThisArgPos then destByrefTy st.oglobals ty else ty
```
Comment (TIHan) [inline src/fsharp/TastPickle.fs]: Why are we getting the destination of the `inref` here?

File: src/fsharp/TastPickle.fs [hunk 4]
```diff
@@ -1277,7 +1287,13 @@ let p_nonlocal_val_ref (nlv:NonLocalValOrMemberRef) st =
     p_bool pkey.MemberIsOverride st 
     p_string pkey.LogicalName st 
     p_int pkey.TotalArgCount st 
+    let isStructThisArgPos = 
+        match key.TypeForLinkage with 
+        | None -> false
+        | Some ty -> checkForInRefStructThisArg st ty
+    st.isStructThisArgPos <- isStructThisArgPos
```
[the remark this answers is not in the document]
Comment (TIHan) [reply src/fsharp/TastPickle.fs]: I agree this looks weird; if it needs to be set back to `false` we should have a comment explaining why.

**Comment by TIHan** [comment]
This looks good to me. I just had a few questions regarding some of the code because I don't know the meaning of a couple things. :)

File: src/fsharp/TastPickle.fs [hunk 5]
```diff
@@ -1547,7 +1564,11 @@ let _ = fill_p_ty (fun ty st ->
               p_byte 0 st; p_tys l st
     | TType_app(ERefNonLocal nleref,[]) -> p_byte 1 st; p_simpletyp nleref st
     | TType_app (tc,tinst)              -> p_byte 2 st; p_tup2 (p_tcref "typ") p_tys (tc,tinst) st
-    | TType_fun (d,r)                   -> p_byte 3 st; p_tup2 p_ty p_ty (d,r) st
+    | TType_fun (d,r)                   -> 
+        p_byte 3 st
+        p_ty d st
+        st.isStructThisArgPos <- false
```
Comment (TIHan) [inline src/fsharp/TastPickle.fs]: Why are we setting this to `false` here?

**Comment by KevinRansom** [comment]
Lol … it's going to ship like this, so take some time on the cleanup.

File: src/fsharp/TastPickle.fs (hunk 1 above)
Comment (dsyme) [reply src/fsharp/TastPickle.fs]: Like .NET metadata, pickled F# metadata contains references from one assembly to another. These must be "linked".  The "TypeForLinkage" is thus the signature of a method that might be overloaded, used to work out which method is being referred to in another assembly.

File: src/fsharp/TastPickle.fs (hunk 2 above)
Comment (dsyme) [reply src/fsharp/TastPickle.fs]: Just a `TcGlobals` like `g` elsewhere in the compiler.

File: src/fsharp/TastPickle.fs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dsyme) [reply src/fsharp/TastPickle.fs]: It's awful. Pushing a different (but equivalent) PR to clean this up

File: src/fsharp/TastPickle.fs (hunk 3 above)
Comment (dsyme) [reply src/fsharp/TastPickle.fs]: Oh crap, yes, this is wrong, It should be mkByrefTy (destByrefTy st.oglobals ty), i.e. convert the `inref` to a `byref`.  

In a way this shows how much the stored `this` type doesn't actually affect things in the compiler, but I'll definitely adjust to get it right.

**Pull request merged**